Wallet support

dkent600 opened the pull request:
Resolves https://daostack.tpondemand.com/RestUI/Board.aspx#page=board/5209716961861964288&appConfig=eyJhY2lkIjoiQjgzMTMzNDczNzlCMUI5QUE0RUE1NUVEOUQyQzdFNkIifQ==&boardPopup=feature/1156/silent

- supports extension wallets (like MetaMask and Safe) as well as Portis and Fortmatic
- Adds Login and Logout buttons in the user account dropdown
- The login button, rather than showing the entire list of providers, reloads cached provider information to automatically log the user into their previously-selected (and therefore cached) provider

**Notes**
- there are lots of permutations here even when not accounting for support for new providers.  Test thorougly!
- see the comment I made below about the login/logout icons

dkent600 commented:
@A-Zak   I think the logout icon you gave me is more appropriate as a login icon, like the metaphor given in these:

Login like this:  
![image](https://user-images.githubusercontent.com/1821666/62224659-05bff800-b385-11e9-9560-2b834124e497.png)


And a logout icon would be like this:

![image](https://user-images.githubusercontent.com/1821666/62224554-e5903900-b384-11e9-9ddd-5de2e1f6d917.png)

leviadam commented:
Maybe one small comment @dkent600.
The login text inside the hover has a MM logo next to it, even when you know it is not a MM account, which is very strange.
Say the storage knows that I'm a Portis user, and it shows my account, but I still have a log in button with a MM logo, it is very confusing.

Either not show a logo there, or show the correct one.

tibetsprague commented on src/layouts/AppContainer.tsx:
```diff
@@ -64,59 +66,88 @@ class AppContainer extends React.Component<IProps, IState> {
   }
 
-  public render() {
+  public render(): any {
```
yeah im over this lint rule

dkent600 replied:
Not me, at all.  'any' is really not a big deal.  Not having the the non-trivial return values is.

leviadam replied:
I think if both @jellegerbrandy and @tibetsprague prefer not to have this rule we should turn it off @dkent600.

dkent600 commented:
@leviadam   @A-Zak 

> Either not show a logo there, or show the correct one.

I totally agree, but it is my understanding that this is Alex's call.

dkent600 commented on src/arc.ts:
```diff
@@ -37,9 +73,58 @@ const settings = {
+/**
+ * Return the default account, locked, in current use by Arc.
+ */
+async function _getCurrentLockedAccount(): Promise<string> {
```
The assumption is that `account[0]` is only available when locked.

jellegerbrandy replied:
but why would it be *locked"? Is that even relevent here? It is a  confsing name, why not rename it "getCurrentAccountFromProvider"?

dkent600 commented on src/arc.ts:
```diff
@@ -101,128 +187,287 @@ export async function checkWeb3Provider() {
-  } else {
-    console.log("Using default Web3 (read-only) provider");
+/**
+ * @return the currently-locked account address from Arc. Ignores any injected
```
Could also be I'm also not using the term correctly.  By "locked" I mean that there is an account that can be used to sign transactions.

tibetsprague replied:
gotcha. where does that term come from?

leviadam replied:
"Locked" does not sound like a good term for that. Active maybe?

jellegerbrandy replied:
we can just drop the adjective. "Current acocunt" instead of "currently-locked account"?

dkent600 commented on src/arc.ts:
```diff
@@ -37,9 +73,58 @@ const settings = {
+      break;
+    }
+    default: {
+      console.log(process.env.NODE_ENV === "development");
```
Done

dkent600 commented on src/arc.ts:
```diff
@@ -1,7 +1,43 @@
+          key: "pk_live_38A2BD2B1D4E9912",
+        },
+      }
+      : // (process.env.NODE_ENV === "staging") ?
```
Yeah, I guess it's OK to hide Portis and Fortmatic when in development mode.

leviadam commented:
> @leviadam @A-Zak
> 
> > Either not show a logo there, or show the correct one.
> 
> I totally agree, but it is my understanding that this is Alex's call.

Just so that we are on the same page here. This is not the same logo of MM we discussed before. Before we talked about the one shown if there is no info, and when you click on it, you get the web3connect options. Now I'm talking about the one inside the hover, in which you already know what is the provider.

dkent600 commented:
@leviadam 

> > > Either not show a logo there, or show the correct one.
> > 
> > I totally agree, but it is my understanding that this is Alex's call.
> Just so that we are on the same page here. This is not the same logo of MM we discussed before. 

See my comments above about the icons.  That is exactly what I was referring to. :-)

jellegerbrandy commented:
I am having some problems with the review app: https://alchemy-staging-rinkeby-pr-854.herokuapp.com

Basically, the Metamask button on the "log in" screen is unresponsive, and I see this in the console:
* 
![image](https://user-images.githubusercontent.com/1306173/62282815-87308c80-b450-11e9-8c38-4827273b7c7d.png)

* 
![image](https://user-images.githubusercontent.com/1306173/62282866-9ca5b680-b450-11e9-84c4-1aeccca7e6cf.png)

jellegerbrandy commented:
Clicking on the fortmatic link gives me this: 
![image](https://user-images.githubusercontent.com/1306173/62283248-69aff280-b451-11e9-94b7-a853297211fe.png)

jellegerbrandy commented:
I also get lots of console errors trying to login to portis (with a non-existing account) that are not reported in the UX: 
![image](https://user-images.githubusercontent.com/1306173/62283475-e9d65800-b451-11e9-8b33-3d5447077477.png)

jellegerbrandy commented:
![image](https://user-images.githubusercontent.com/1306173/62283673-594c4780-b452-11e9-88bb-85bd01859641.png)

dkent600 commented:
@jellegerbrandy

The unique key errors exist in both Alchemy and Web3Connect.  I have already created tickets on both these errors in the respective repos.

dkent600 commented:
> the Metamask button on the "log in" screen is unresponsive

It works fine for me.  I could use a repro case on this (steps to reproduce).

A-Zak commented:
@dkent600 let's use the icons you found for login and logout...
and for the main login button, let's just remove the MM icon...

dkent600 commented:
@jellegerbrandy 

> Clicking on the fortmatic link gives me this:

Fortmatic works fine for me.  Do you have a fortmatic account?

> I also get lots of console errors trying to login to portis (with a non-existing account)

Same for Portis, no problem for me.

I think I need better repro cases here, like the steps that occur when there is no account.

dkent600 commented:
@jellegerbrandy 

> the Metamask button on the "log in" screen is unresponsive

I am guessing you are connected to the wrong network, and we aren't showing the error message except in the console. I will fix this.

dkent600 commented:
@jellegerbrandy I have done some refactoring that cleans up some code and improves the logging and error handling/reporting.  If your problem with MM was due to being connected to the wrong network, you should now see a notification to that effect.

dkent600 commented:
@jellegerbrandy @A-Zak @leviadam @tibetsprague   With today's dev meeting conversation in mind (about letting the user know a little more about what is happening), I added notifications for when logging in and lgging out.

Also updated the icons.

dkent600 commented:
@jellegerbrandy  I am not having any problem with Portis nor Fortmatic with no account (though honestly I'm not sure what you mean by that, I just used incognito mode  -- need a fleshed-out repro case)
